docs/content/envoy-primer.md

```diff
@@ -0,0 +1,347 @@
+
+```live:bool_example:module:read_only
+package envoy.authz
+
+import input.attributes.request.http as http_request
+
+default allow = false
+
+token = {"valid": valid, "payload": payload} {
+    [_, encoded] := split(http_request.headers.authorization, " ")
+    [valid, _, payload] := io.jwt.decode_verify(encoded, {"secret": "secret"})
+}
+
+allow {
+    is_token_valid
+    action_allowed
+}
+
+is_token_valid {
+  token.valid
+  now := time.now_ns() / 1000000000
+  token.payload.nbf <= now
+  now < token.payload.exp
+}
+
+action_allowed {
+  http_request.method == "GET"
+  token.payload.role == "guest"
+  glob.match("/people*", [], http_request.path)
+}
+
+action_allowed {
+  http_request.method == "GET"
+  token.payload.role == "admin"
+  glob.match("/people*", [], http_request.path)
+}
+
+action_allowed {
+  http_request.method == "POST"
+  token.payload.role == "admin"
+  glob.match("/people", [], http_request.path)
+  lower(input.parsed_body.firstname) != base64url.decode(token.payload.sub)
+}
+```
+
+The first line `package envoy.authz` declaration gives the (hierarchical) name `envoy.authz` to the rules in the
+remainder of the policy. If the OPA-Envoy [configuration](envoy-introduction#configuration) does not specify the `path`
+field, `envoy/authz/allow` will be considered as the default policy decision path. `data.envoy.authz.allow` will be the
+name of the policy decision to query in the default case.
+
+The line `import input.attributes.request.http as http_request`, defines a dependency on `attributes.request.http`
+which is part of the [input document](#input-document). The line also uses the `as` keyword so that `http_request` can
+be used as an alias for `input.attributes.request.http` in the rest of the policy.
+
+The above policy uses the `io.jwt.decode_verify` builtin function to parse and verify the JWT containing
+information about the user making the request. It uses other builtins like `glob.match`, `lower`, `base64url.decode` etc.
+OPA has 150+ builtins detailed at [openpolicyagent.org/docs/policy-reference](../policy-reference).
+
+The dot notation seen in multiple places in the policy for ex. `input.parsed_body.firstname` simply descends through
+the hierarchy to access the requested value. The dot (.) operator never throws any errors; if the path does not exist
+the value of the expression is `undefined`.
+
+### Example Policy with Object Response
+
+The `allow` rule in the above policy returns a `boolean` decision to indicate whether a request should be allowed or not.
+If you'd like your rule to not only indicate if a request is allowed or not but also provide optional response headers,
+body and HTTP status that can be sent to the downstream client or upstream, the below `allow` rule generates an `object`
+that provides additional details along with the status of the request (ie. `allowed` or `denied`).
+
+```live:obj_example:module:read_only
+package envoy.authz
+
+default allow = {
```

Comment:
I'm on the fence about whether we should call this top-level rule `allow` or not. On the other hand, if we change the rule, we need to be very clear to tell the user to change the configuration.